src/gui/guiScene.cpp

```diff
@@ -0,0 +1,262 @@
+/*
+Minetest
+Copyright (C) 2020 Jean-Patrick Guerrero <jeanpatrick.guerrero@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "guiScene.h"
+
+#include <SViewFrustum.h>
+#include <IAnimatedMeshSceneNode.h>
+#include <ILightSceneNode.h>
+#include "porting.h"
+
+GUIScene::GUIScene(gui::IGUIEnvironment *env, scene::ISceneManager *smgr,
+		   gui::IGUIElement *parent, core::recti rect, s32 id)
+	: IGUIElement(gui::EGUIET_ELEMENT, env, parent, id, rect)
+{
+	m_driver = env->getVideoDriver();
+	m_smgr = smgr->createNewSceneManager(false);
+
+	m_cam = m_smgr->addCameraSceneNode(0, v3f(0.f, 0.f, -100.f), v3f(0.f));
+	m_cam->setFOV(30.f * core::DEGTORAD);
+
+	scene::ILightSceneNode *light = m_smgr->addLightSceneNode(m_cam);
+	light->setRadius(1000.f);
+
+	m_smgr->getParameters()->setAttribute(scene::ALLOW_ZWRITE_ON_TRANSPARENT, true);
+}
+
+GUIScene::~GUIScene()
+{
+	setMesh(nullptr);
+
+	m_smgr->drop();
+}
+
+scene::IAnimatedMeshSceneNode *GUIScene::setMesh(scene::IAnimatedMesh *mesh)
+{
+	if (m_mesh) {
+		m_mesh->remove();
+		m_mesh = nullptr;
+	}
+
+	if (!mesh)
+		return nullptr;
+
+	m_mesh = m_smgr->addAnimatedMeshSceneNode(mesh);
+	m_mesh->setPosition(-m_mesh->getBoundingBox().getCenter());
+	m_mesh->animateJoints();
+	return m_mesh;
+}
+
+void GUIScene::setTexture(u32 idx, video::ITexture *texture)
+{
+	video::SMaterial &material = m_mesh->getMaterial(idx);
+	material.MaterialType = video::EMT_TRANSPARENT_ALPHA_CHANNEL;
+	material.MaterialTypeParam = 0.5f;
+	material.TextureLayer[0].Texture = texture;
+	material.setFlag(video::EMF_LIGHTING, false);
+	material.setFlag(video::EMF_FOG_ENABLE, true);
+	material.setFlag(video::EMF_BILINEAR_FILTER, false);
+	material.setFlag(video::EMF_BACK_FACE_CULLING, false);
+}
+
+void GUIScene::draw()
+{
+	// Control rotation speed based on time
+	u64 new_time = porting::getTimeMs();
+	u64 dtime_ms = 0;
+	if (m_last_time != 0)
+		dtime_ms = porting::getDeltaMs(m_last_time, new_time);
+	m_last_time = new_time;
+
+	core::rect<s32> oldViewPort = m_driver->getViewPort();
+	m_driver->setViewPort(getAbsoluteClippingRect());
```

Comment:
This causes some unintended size reduction of the model if element clipped. It could be weird in scrollable containers.
Scene should be drawn in a view port sized as `AbsoluteRect` and clipped to `getAbsoluteClippingRect()`, a bit like in `drawItemStack`.

Reply:
Could you please update the code according to your plans? I don't know what to do there.

Reply:
Seems a bit more complicated than I thought. I'm having a look.